framework/src/physics/PhysicsBase.C

```diff
@@ -0,0 +1,254 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PhysicsBase.h"
+#include "MooseUtils.h"
+#include "FEProblemBase.h"
+
+#include "NonlinearSystemBase.h"
+#include "AuxiliarySystem.h"
+#include "BlockRestrictable.h"
+
+InputParameters
+PhysicsBase::validParams()
+{
+  InputParameters params = Action::validParams();
+  params.addClassDescription("Creates all the objects necessary to solve a particular physics");
+
+  params.addParam<std::vector<SubdomainName>>(
+      "block", {}, "Blocks that this Physics is active on. Components can add additional blocks");
+
+  MooseEnum transient_options("true false same_as_problem", "same_as_problem");
+  params.addParam<MooseEnum>(
+      "transient", transient_options, "Whether the physics is to be solved as a transient");
+
+  MooseEnum pc_options("default none", "none");
+  params.addParam<MooseEnum>(
+      "preconditioning", pc_options, "Which preconditioning to use for this Physics");
+  params.addParam<bool>("verbose", false, "Flag to facilitate debugging a Physics");
+
+  // Restart parameters
+  params.addParam<bool>("initialize_variables_from_mesh_file",
+                        false,
+                        "Determines if the variables that are added by the action are initialized"
+                        "from the mesh file (only for Exodus format)");
+  params.addParam<std::string>(
+      "initial_from_file_timestep",
+      "LATEST",
+      "Gives the time step number (or \"LATEST\") for which to read the Exodus solution");
+  params.addParamNamesToGroup("initialize_variables_from_mesh_file initial_from_file_timestep",
+                              "Restart from Exodus");
+  return params;
+}
+
+PhysicsBase::PhysicsBase(const InputParameters & parameters)
+  : Action(parameters),
+    _sys_number(0),
+    _verbose(getParam<bool>("verbose")),
+    _preconditioning(getParam<MooseEnum>("preconditioning")),
+    _blocks(getParam<std::vector<SubdomainName>>("block")),
+    _is_transient(getParam<MooseEnum>("transient"))
+{
+  checkSecondParamSetOnlyIfFirstOneTrue("initialize_variables_from_mesh_file",
+                                        "initial_from_file_timestep");
+  prepareCopyNodalVariables();
+  addRequiredPhysicsTask("init_physics");
+}
+
+void
+PhysicsBase::act()
+{
+  mooseDoOnce(checkRequiredTasks());
+
+  if (_current_task == "init_physics")
```

Comment:
@loganharbour 

this is the first big change between physics and actions.
We'd expect Physics developers to implement the relevant routines and register the tasks, instead of re-writing some version of this act() routine

this act() looks a lot like other physics-defining actions (TM, NS)